[Event Hubs] Buffered Producer Live Test and Fixes

jsquire opened the pull request:
# Summary

The focus of these changes is to implement the live tests for end-to-end scenarios.  Where the tests exposed flaws, they were corrected and a general polish pass was done to the implementation.  This set of changes also makes the producer and associated support types `public` in anticipation of the pending beta release.  Also riding along are some minor formatting changes.

Pending work:
- Samples

jsquire committed:
[Event Hubs] Buffered Producer Live Test and Fixes

The focus of these changes is to implement the live tests for end-to-end
scenarios. Where the tests exposed flaws, they were corrected and a general
polish pass was done to the implementation.

This set of changes also makes the producer and associated support types
public in anticipation of the pending beta release. Also riding along are
some minor formatting changes.

jsquire commented:
/azp run net - eventhub - tests

azure-sdk commented:
API changes have been detected in `Azure.Messaging.EventHubs`. You can review API changes [here](https://apiview.dev/Assemblies/Review/300db86d834d4bb681f564357abf25cf?diffOnly=True&diffRevisionId=2c4ed251b712451088a66054659e43e7)

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:
```diff
@@ -110,6 +111,9 @@ internal class EventHubBufferedProducerClient : IAsyncDisposable
         /// <summary>The count of total events that have been buffered across all partitions.</summary>
         private int _totalBufferedEventCount;
 
+        /// <summary>The set of publishing event handlers that are actively executing; these should be awaited when closing. </summary>
```
```suggestion
        /// <summary>The set of publishing event handlers that are actively executing; these should be awaited when flushing. </summary>
```

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:
```diff
@@ -1350,9 +1380,9 @@ internal virtual async Task FlushInternalAsync(CancellationToken cancellationTok
-            var batchEventCount = 0;
             var operationId = GenerateOperationId();
             var partitionId = partitionState.PartitionId;
+            var eventsInBatch = new List<EventData>();
```
Rename to `batchEvents` to match other methods.

christothes commented on sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs:
```diff
@@ -2151,6 +2221,39 @@ private async Task RunProducerManagementAsync(CancellationToken cancellationToke
+                // are used for different purposes, it is permissible for them to drift for a short time.
+                // As a result, there's no need to synchronize them.
+
+                var queriedHash = new HashSet<string>(queriedPartitions);
```
nit: is there any value in storing this as a local variable rather than just assigning it to `_partitionHash`?

jsquire committed:
Addressing Chris' feedback, ensuring that continuations for `TaskCompletionSource`
instances used in tests are run asynchronously, and updating the summary comments
for the `EventHubProducerClient` to reference the option of using the buffered
producer to reduce complexity.

jsquire commented:
/azp run net - eventhub - tests

jsquire commented:
Test failure was an unrelated intermittent; moving forward with the merge.

jsquire commented:
/check-enforcer override
